lib/logstash/inputs/snmptrap.rb

```diff
@@ -47,26 +47,44 @@ def register
     end
   end # def register
 
-  public
   def run(output_queue)
     begin
       # snmp trap server
       snmptrap_listener(output_queue)
     rescue => e
       @logger.warn("SNMP Trap listener died", :exception => e, :backtrace => e.backtrace)
-      sleep(5)
-      retry
+      stoppable_call(20) { sleep(0.25) }
```

Comment:
For my part I don't like the way the pipeline holds unrelated references to the thread and the plugin in different arrays.  I would consider a tuple value object PluginRunner(?), with the thread and instance as tuple values and this instance being stored in one array.
```ruby
def start_input
  thread = Thread.new { inputworker(plugin) }
  @input_runners << PluginRunner.new(input, thread)
end

def shutdown
  #...
  @input_runners.each do |runner|
    runner.plugin.stop
    runner.thread.run #noop if thread is running
  end
end
```
Obviously the names of ivars, variables and classes should be appropriate.
I'm thinking this should work for thread sleep called from Stud as well?